lib/shared/bucketing-assembly-script/assembly/eventQueue/requestPayloadManager.ts

```diff
@@ -101,32 +101,59 @@ export class RequestPayloadManager {
             }
         }
 
-        // Generate defaulted agreagate user as our events APIs require a user / user_id
+        // Generate defaulted aggregate user as our events APIs require a user / user_id
         const dvcUser = new DVCPopulatedUser(new DVCUser(
             user_id, null, null, null, null, NaN, null, null, null, null
         ))
         this.addEventsToPendingPayloads(new UserEventsBatchRecord(dvcUser, aggEvents))
     }
 
+    /**
+     * Find existing FlushPayload that contains records for the user_id,
+     * or return the smallest FlushPayload or a new FlushPayload
+     */
+    private getFlushPayloadForUser_id(user_id: string): FlushPayload {
+        const keys = this.pendingPayloads.keys()
+        let smallestPayload: FlushPayload | null = null
+
+        for (let i=0; i < keys.length; i++) {
+            const key = keys[i]
+            const payload = this.pendingPayloads.get(key)
+            const recordForUser = payload.recordForUser(user_id)
+            const payloadEventCount = payload.eventCount()
+            if (recordForUser && payloadEventCount < this.chunkSize) {
+                this.pendingPayloads.delete(key)
+                return payload
+            }
+            if (payloadEventCount < this.chunkSize
+                && (!smallestPayload || smallestPayload.eventCount() < payloadEventCount)) {
+                smallestPayload = payload
+            }
+        }
+
+        if (smallestPayload) {
+            this.pendingPayloads.delete(smallestPayload.payloadId)
+            return smallestPayload
+        } else {
+            return new FlushPayload([])
+        }
+    }
+
     /**
      * Chunk up UserEventsBatchRecord's into payloads of size: this.chunkSize
      */
     private addEventsToPendingPayloads(record: UserEventsBatchRecord): void {
-        let flushPayload = new FlushPayload([])
+        let flushPayload = this.getFlushPayloadForUser_id(record.user.user_id)
```

Comment:
personally I think the flush payloads should be immutable once created. What you're doing here potentially adds more events to a payload that was already sent and then failed. Once a payload is returned to native code I dont think it should be possible for it to change.

Reply:
yea I can make an exception for `failed` payloads if we want, however the reason this is needed is because user events and aggregated events are added to the payloads seperatly. Which is what was resulting in the system creating two payloads every time you logged a user event an aggVariableAggregated: https://github.com/DevCycleHQ/js-sdks/pull/248/files/79a82abe111f7aa19e7c1c58b8691c4acee419c6#diff-5e909c6d5ed133d11580490d906b2f23a686d4e7e83e34ddac71713eb3100dc3R26